cpp/src/knn/knn_opg_common.cuh

```diff
@@ -68,69 +68,217 @@ namespace opg {
+/**
+ * The enumeration of KNN distributed operations
+ */
+enum knn_operation {
+  knn,            /**< Simple KNN */
+  classification, /**< KNN classification */
+  class_proba,    /**< KNN classification probabilities */
+  regression      /**< KNN regression */
+};
+
+/**
+ * A structure to store parameters for distributed KNN
+ */
+struct opg_knn_param {
+  union labels_data {
+    std::vector<std::vector<int *>> *i;
+    std::vector<std::vector<float *>> *f;
+  };
+
+  union outputs_data {
+    std::vector<Matrix::Data<int> *> *i;
+    std::vector<Matrix::Data<float> *> *f;
+  };
+
+  knn_operation knn_op; /**< Type of KNN distributed operation */
+  std::vector<Matrix::Data<int64_t> *> *out_I =
+    nullptr; /**< KNN indices output array */
+  std::vector<Matrix::floatData_t *> *out_D =
+    nullptr; /**< KNN distances output array */
+  std::vector<Matrix::floatData_t *> *idx_data =
+    nullptr; /**< Index input array */
+  Matrix::PartDescriptor *idx_desc =
+    nullptr; /**< Descriptor for index input array */
+  std::vector<Matrix::floatData_t *> *query_data =
+    nullptr; /**< Query input array */
+  Matrix::PartDescriptor *query_desc =
+    nullptr;             /**< Descriptor for query input array */
+  bool rowMajorIndex;    /**< Is index row major? */
+  bool rowMajorQuery;    /**< Is query row major? */
+  size_t k = 0;          /**< Number of nearest neighbors */
+  size_t batch_size = 0; /**< Batch size */
+  bool verbose;          /**< verbose */
+
+  labels_data y;     /**< Labels input array (cl&re) */
+  int n_outputs = 0; /**< Number of outputs per query (cl&re) */
+  outputs_data out;  /**< KNN outputs output array (cl&re) */
+
+  std::vector<int *> *uniq_labels =
+    nullptr; /**< Unique labels (classification) */
+  std::vector<int> *n_unique =
+    nullptr; /**< Number of unique labels (classification) */
+
+  std::vector<std::vector<float *>> *probas =
+    nullptr; /**< KNN classification probabilities output array (class-probas) */
+};
+
+/**
+ * A structure to store utilities for CUDA and RAFT comms
+ */
+struct cuda_utils {
+  cuda_utils(raft::handle_t &handle) {
+    this->alloc = handle.get_device_allocator();
+    this->stream = handle.get_stream();
+    this->comm = &(handle.get_comms());  //communicator_ is a private attribute
+    size_t n_internal_streams = handle.get_num_internal_streams();
+    this->internal_streams.resize(n_internal_streams);
+    for (int i = 0; i < n_internal_streams; i++) {
+      internal_streams[i] = handle.get_internal_stream(i);
+    }
+  };
+  std::shared_ptr<deviceAllocator> alloc; /**< RMM alloc */
+  cudaStream_t stream;                    /**< CUDA user stream */
+  const raft::comms::comms_t *comm;       /**< RAFT comms handle */
+  std::vector<cudaStream_t>
+    internal_streams; /**< Vector of CUDA internal streams */
+};
+
+/**
+ * A structure to store utilities for distributed KNN operations
+ */
+struct opg_knn_utils {
```

Comment:
Nitpick: When I think of `utils`, I generally think of statically available functions that are used in many different places in the code but don't necessarily provide enough functionality to be useful as their own class. This seems to be a class for encapsulating state that can be passed around (e.g. POCO object).